src/tests.rs

```diff
@@ -106,6 +106,89 @@ macro_rules! nodes {
     }};
 }
 
+#[derive(Debug)]
+enum MergeStateTag {
+    Local,
+    LocalWithNewLocalStructure,
+    Remote,
+    RemoteWithNewRemoteStructure,
+    Unchanged,
+    UnchangedWithNewLocalStructure,
+}
+
+#[derive(Debug)]
+struct TestMergedNode {
+    guid: Guid,
+    state: MergeStateTag,
+    children: Vec<TestMergedNode>,
+}
+
+impl TestMergedNode {
+    fn new(guid: Guid, state: MergeStateTag) -> TestMergedNode {
+        TestMergedNode {
+            guid,
+            state,
+            children: Vec::new(),
+        }
+    }
+}
+
+impl<'t> PartialEq<MergedNode<'t>> for TestMergedNode {
+    fn eq(&self, other: &MergedNode<'t>) -> bool {
+        if self.guid != other.guid {
+            return false;
+        }
+        let merge_state_matches = match (&self.state, other.merge_state) {
+            (MergeStateTag::Local, MergeState::LocalOnly(_)) => true,
+            (
+                MergeStateTag::LocalWithNewLocalStructure,
+                MergeState::LocalOnlyWithNewLocalStructure(_),
+            ) => true,
+            (MergeStateTag::Remote, MergeState::RemoteOnly(_)) => true,
+            (
+                MergeStateTag::RemoteWithNewRemoteStructure,
+                MergeState::RemoteOnlyWithNewRemoteStructure(_),
+            ) => true,
+            (MergeStateTag::Local, MergeState::Local { .. }) => true,
+            (
+                MergeStateTag::LocalWithNewLocalStructure,
+                MergeState::LocalWithNewLocalStructure { .. },
+            ) => true,
+            (MergeStateTag::Remote, MergeState::Remote { .. }) => true,
+            (
+                MergeStateTag::RemoteWithNewRemoteStructure,
+                MergeState::RemoteWithNewRemoteStructure { .. },
+            ) => true,
+            (MergeStateTag::Unchanged, MergeState::Unchanged { .. }) => true,
+            (
+                MergeStateTag::UnchangedWithNewLocalStructure,
+                MergeState::UnchangedWithNewLocalStructure { .. },
+            ) => true,
+            _ => false,
+        };
+        if !merge_state_matches {
+            return false;
+        }
+        self.children == other.merged_children
+    }
+}
+
+macro_rules! merged_nodes {
```

Comment:
(All the changes to this file are converting `expected_tree` to use `merged_nodes`, so we can check the merge state. Before, we'd only check if the nodes were flagged for upload, but now that we have `LocalWithNewLocalStructure`, it's important we check if we need to apply the node, too).